corehq/apps/accounting/tests/forms/test_mocked_invoice_form.py

```diff
@@ -0,0 +1,124 @@
+
+
+def create_form(domain='test_domain', month=1, year=2020, num_users=0, show_testing_options=False):
+    data = {
+        'month': month,
+        'year': year,
+        'domain': domain,
+        'num_users': num_users
+    }
+    return TriggerInvoiceForm(data, show_testing_options=show_testing_options)
+
+
+class TestTriggerInvoiceFormValidity(SimpleTestCase):
+    def test_has_all_details_is_valid_form(self):
+        form = create_form()
+        self.assertTrue(form.is_valid())
+
+    def test_missing_month_is_invalid(self):
+        form = create_form(month=None)
+        self.assertFalse(form.is_valid())
+
+    def test_missing_year_is_invalid(self):
+        form = create_form(year=None)
+        self.assertFalse(form.is_valid())
+
+    def test_missing_domain_is_invalid(self):
+        form = create_form(domain='')
+        self.assertFalse(form.is_valid())
+
+    def test_current_periods_are_not_valid(self):
+        current_date = datetime.date(2020, 1, 10)
+        form = create_form(month=1, year=2020)
+        with mock.patch('corehq.apps.accounting.forms.datetime') as mock_datetime:
+            mock_datetime.date.today.return_value = current_date
+            self.assertFalse(form.is_valid())
+
+    def test_if_testing_options_are_enabled_then_num_users_is_displayed(self):
+        form = create_form(show_testing_options=True)
+        self.assertIn('num_users', form.fields)
+
+    def test_if_testing_options_are_disabled_then_num_users_is_hidden(self):
+        form = create_form(show_testing_options=False)
+        self.assertNotIn('num_users', form.fields)
+
+
+# cannot use SimpleTestCase due to transaction.atomic -- and that behavior should be tested.
+# Ultimately it likely means form and database behavior should be separated,
+# so that form can use SimpleTestCase, and database can use TestCase
+class TestMockedInvoiceForm(TestCase):
+    def setUp(self):
+        self.month = 1
+        self.year = 2020
+        self.start_date = datetime.date(self.year, self.month, 1)
+        self.end_date = datetime.date(self.year, self.month, 31)
+
+        self.create_mocks()
+
+    def create_mocks(self):
+        get_invoices_patcher = patch.object(Invoice,
+                'get_domain_invoices_between_dates', return_value=[])
+        self.get_existing_invoices = get_invoices_patcher.start()
+        self.addCleanup(get_invoices_patcher.stop)
+
+        invoice_factory_patcher = patch('corehq.apps.accounting.forms.DomainInvoiceFactory')
+        invoice_factory_patcher.start()
+        self.addCleanup(invoice_factory_patcher.stop)
+
+        # Reverse mocked out because it takes ~1 second to generate the lookup table
+        reverse_patcher = mock.patch('corehq.apps.accounting.forms.reverse',
+            return_value='some.url')
+        reverse_patcher.start()
+        self.addCleanup(reverse_patcher.stop)
+
+    def create_form(self, **kwargs):
+        kwargs['month'] = self.month
+        kwargs['year'] = self.year
```

Comment:
I don't think my current version is particularly more readable than yours, but it would basically be a violation of YAGNI -- my tests don't currently override those parameters, so there isn't a good reason to allow them to be overridden yet.